### PR TITLE
virt: Initial implementation using zydis and xbyak

### PR DESCRIPTION
**Pull request by skmp**
### Overview
This is a cut-down version of #23, meant to provide very basic, simple virtualization at the cost of perfomance.

### Goals
- Simple implementation
- Virtualize all registers
- No register allocation
- Support most (all?) ps4 instructions/regs, though just passthrough
- Basic, per thread code cache
- No SMC support
- No instruction rewriting for most operations

### Todo
- [x] Get basic hack-and-slash demo working
- [ ] Decide on a testing stategy
- [ ] Rework to less horrific code
- [ ] Support multiple threads, TCB overflows, and other corner cases gracefully

### CODE REVIEW

**Comment by skmp** [comment]
Gets to the first call now!

**Comment by georgemoralis** [comment]
;0

**Comment by skmp** [comment]
video out test will now run, glitchy.

Screenshot 2023-10-23 012140

Also tetris, and openbor gets quite far (crashes at some vsnprintf with null format string)

File: src/Core/PS4/Linker.cpp [hunk 1]
```diff
@@ -593,6 +597,21 @@ void Linker::Relocate(Module* m)
 	}
 }
 
+void GenerateTrampoline(u64 hle_handler, u64 context_base);
+
+__declspec(align(32)) struct Context {
```
Comment (raphaelthegreat) [inline src/Core/PS4/Linker.cpp]: Can this use alignas(32) as more portable alternative?

File: src/Core/PS4/Linker.cpp [hunk 2]
```diff
@@ -650,37 +669,561 @@ static PS4_SYSV_ABI void ProgramExitFunc() {
     printf("exit function called\n");
 }
 
+auto gen = new Xbyak::CodeGenerator(64 * 1024 * 1024);
```
Comment (raphaelthegreat) [inline src/Core/PS4/Linker.cpp]: std::make_unique

File: src/Core/PS4/Linker.cpp [hunk 3]
```diff
@@ -650,37 +669,561 @@ static PS4_SYSV_ABI void ProgramExitFunc() {
...
+
+
+constexpr auto rip_offs = offsetof(Context, rip);
+constexpr auto ymm_offs = offsetof(Context, ymm[0]);
+constexpr auto ymm_size = 32;
+constexpr auto rflags_offs = offsetof(Context, rflags);
+constexpr auto host_rsp_offs = offsetof(Context, host_rsp);
+constexpr auto trampoline_ret_offs = offsetof(Context, trampoline_ret);
+
+std::unordered_map<u64, u64 (*)()> translated_entries;
```
Comment (raphaelthegreat) [inline src/Core/PS4/Linker.cpp]: Better mark this as static if it's not going to be used in other cpp files

**Review by raphaelthegreat — COMMENTED** [review]
From an architecture perspective, I feel generation could be simplified by making a class that inherits from xbyak::CodeGenerator. Also you can better control the generator lifetime instead of it being in global scope.

**Comment by georgemoralis** [comment]
Vnspritnf use the same algorithm as printf. So if it crashes That will crash as well

**Comment by georgemoralis** [comment]
Printf is special implementation since sysv abi can't parse va-list for arguments

**Comment by skmp** [comment]
@GPUCode this is more or less a hack-and-slash for now, more to explore what needs to be done and less to be merged, for now. The code is also quite horrific as is, it will be rewritten before merging.

A couple of notes so far
- zydis + xbyak isn't a great combination
- x86 decoding, as always, is a huge gotcha within a gotcha
- segment overrides need to be handled
- I have no plan on how to handle multi threaded contexts for now

**Comment by skmp** [comment]
video out sample works now, as well as openbor, and also release builds.

File: src/Core/PS4/Linker.cpp [hunk 4]
```diff
@@ -650,37 +669,556 @@ static PS4_SYSV_ABI void ProgramExitFunc() {
...
 
-        "pushq 8(%1)\n"		// copy EntryParams to top of stack like the kernel does
-        "pushq 0(%1)\n"		// OpenOrbis expects to find it there
+        rsp = rsp & ~16;
+        rsp = rsp - 8;
 
-		"movq %1, %%rdi\n"	// also pass params and exit func
-        "movq %2, %%rsi\n"	// as before
+        rsp = rsp - 8;
+        *(void**)rsp = params->argv;
```
Comment (red-prig) [inline src/Core/PS4/Linker.cpp]: In this case, it is not a link to the argv array that should be passed, but the array of argv links itself directly. So the size will be variable.

**Comment by raphaelthegreat** [comment]
> @GPUCode this is more or less a hack-and-slash for now, more to explore what needs to be done and less to be merged, for now. The code is also quite horrific as is, it will be rewritten before merging.
> 
> A couple of notes so far
> 
>     * zydis + xbyak isn't a great combination
> 
>     * x86 decoding, as always, is a huge gotcha within a gotcha
> 
>     * segment overrides need to be handled
> 
>     * I have no plan on how to handle multi threaded contexts for now

Noted, then better leave these notes for future refactor work :p

File: src/Core/PS4/Linker.cpp (hunk 4 above)
Comment (skmp) [reply src/Core/PS4/Linker.cpp]: Yup, can confirm that the array should be inline there from the assembly
```
TranslateCode: 00000009000B0D18, ctx: 7FF613C59720
00000009000B0D18  xor rbp, rbp
00000009000B0D1B  mov rdi, rsp
00000009000B0D1E  lea rsi, [0x000000090010A628
00000009000B0D25  and rsp, 0xFFFFFFFFFFFFFFF0
00000009000B0D29  call 0x00000009000B0D30
TranslateCode: 00000009000B0D30, ctx: 7FF613C59720
00000009000B0D30  mov esi, [rdi]
00000009000B0D32  lea rdx, [rdi+0x08]
```
Funny how that doesn't match with the openorbis source code, must be looking at the wrong place? Hmm.

Comment (skmp) [reply src/Core/PS4/Linker.cpp]: with `long* p` Open Orbis uses ```	char **argv = (void *)(p+1);```, seems I missread that last time. Thanks @red-prig :)

**Comment by skmp** [comment]
(Rebased over main, squashed a bit)

To cleanup and merge to main this needs
- [ ] Reorganized out of linker
- [ ] Handling of code buffer overflows
- [ ] Separation between Trampoline Generation and linker
- [ ] Separation of global state and per thread state
- [ ] Cleanup of the codegen code, so it is bearable
- [ ] Support for multiple threads

**Comment by georgemoralis** [comment]
quite old and deprecated will check again sometime
